firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java

```diff
@@ -391,10 +376,9 @@ public SnapshotVersion getLastRemoteSnapshotVersion() {
             MaybeDocument existingDoc = existingDocs.get(key);
 
             if (existingDoc == null
-                || (authoritativeUpdates.contains(doc.getKey()) && !existingDoc.hasPendingWrites())
-                || doc.getVersion().compareTo(existingDoc.getVersion()) >= 0) {
-              // If a document update isn't authoritative, make sure we don't apply an old document
-              // version to the remote cache.
+                || doc.getVersion().compareTo(existingDoc.getVersion()) > 0
+                || (doc.getVersion().compareTo(existingDoc.getVersion()) == 0
+                    && existingDoc.hasPendingWrites())) {
```

Comment:
Hmm, why do we overwrite the existing doc when it has pending writes?

Reply:
This is the only important question, the rest are nitpicks.

Reply:
When the write stream sends us an acknowledgment, we persist a document with `hasCommittedMutations`. We then pretend we know the state of the document, but that is not always true (another client may have changed a field we didn't write to, or we supplied an ArrayUnion/ArrayRemove transform that was applied to a set of elements that the client didn't yet know about). By allowing the server to send us the state of a committed document, we overwrite the `hasCommittedMutations` version with the actual version from the backend.

Reply:
Thanks for the explanation!